api/src/main/java/org/openmrs/module/sync2/api/utils/SyncHashcodeUtils.java

```diff
@@ -0,0 +1,143 @@
+import org.openmrs.Location;
+import org.openmrs.Obs;
+import org.openmrs.Order;
+import org.openmrs.Patient;
+import org.openmrs.PatientProgram;
+import org.openmrs.Person;
+import org.openmrs.Privilege;
+import org.openmrs.Program;
+import org.openmrs.Provider;
+import org.openmrs.Relationship;
+import org.openmrs.User;
+import org.openmrs.Visit;
+import org.openmrs.VisitType;
+import org.openmrs.module.sync2.api.model.audit.AuditMessage;
+import org.openmrs.module.sync2.client.SimpleObjectMessageConverter;
+import org.openmrs.module.webservices.rest.SimpleObject;
+
+import java.util.Map;
+import java.util.UUID;
+
+public class SyncHashcodeUtils {
+
+	private static final String LOCATION_ID = "locationId";
+	private static final String OBS_ID = "obsId";
+	private static final String ENCOUNTER_ID = "encounterId";
+	private static final String VISIT_ID = "visitId";
+	private static final String PROVIDER_ID = "providerId";
+	private static final String DRUG_ID = "drugId";
+	private static final String FORM_ID = "formId";
+	private static final String ORDER_ID = "orderId";
+	private static final String PATIENT_PROGRAM_ID = "patientProgramId";
+	private static final String PERSON_ID = "personId";
+	private static final String PATIENT_ID = "patientId";
+	private static final String RELATIONSHIP_ID = "relationshipId";
+	private static final String COHORT_ID = "cohortId";
+	private static final String VISIT_TYPE_ID = "visitTypeId";
+	private static final String USER_ID = "userId";
+	private static final String PROGRAM_ID = "programId";
+	private static final String AUDIT_MESSAGE_ID = "id";
+
+	private static final SimpleObjectMessageConverter converter = new SimpleObjectMessageConverter();
+
+	private static final String regex = "(\\w{8})(\\w{4})(\\w{4})(\\w{4})(\\w{12})";
+
+	private static final String replacement = "$1-$2-$3-$4-$5";
+
+	public static UUID getHashcode(SimpleObject simpleObject, Class<?> clazz) {
+		return getUuidFromString(converter.convertToJson(removeFields(simpleObject, clazz)));
+	}
+
+	private static SimpleObject copySimpleObject(SimpleObject simpleObject) {
+		SimpleObject result = new SimpleObject();
+		for (Map.Entry<String, Object> entry : simpleObject.entrySet()) {
+			result.add(entry.getKey(), entry.getValue());
+		}
+		return result;
+	}
+
+	private static UUID getUuidFromString(String data) {
+		String md5Hex = DigestUtils.md5Hex(data);
+		md5Hex = md5Hex.replaceFirst(regex, replacement);
+		return UUID.fromString(md5Hex);
+	}
+
+	private static SimpleObject removeFields(SimpleObject simpleObject, Class<?> clazz) {
+		SimpleObject result = copySimpleObject(simpleObject);
+		result.remove("dateChanged");
+		removeVoided(result, clazz);
+		removeId(result, clazz);
+		return result;
+	}
+
+	private static void removeVoided(final SimpleObject result, Class<?> clazz) {
+		if (BaseOpenmrsData.class.isAssignableFrom(clazz)) {
+			result.remove("dateVoided");
+			result.remove("voided");
+			result.remove("voidedBy");
```

Comment:
Do we want to delete this information?

Reply:
We have to detect if an object was voided on our instance, so we need to be able to tell if the received object was voided on the current instance. Adding voided to hashcode produce different hashcodes for objects with the same values.